tests/stdlib/tjsonutils.nim

```diff
@@ -4,15 +4,38 @@ discard """
 
 import std/jsonutils
 import std/json
+from std/fenv import epsilon
 
 proc testRoundtrip[T](t: T, expected: string) =
+  # checks that `T => json => T2 => json2` is such that json2 = json
   let j = t.toJson
   doAssert $j == expected, $j
   doAssert j.jsonTo(T).toJson == j
   var t2: T
   t2.fromJson(j)
   doAssert t2.toJson == j
 
+proc testRoundtripVal[T](t: T, expected: string) =
+  # similar to testRoundtrip, but also checks that the `T => json => T2` is such that `T2 == T`
+  # note that this isn't always possible, e.g. for pointer-like types.
+  let j = t.toJson
+  let j2 = $j
+  doAssert j2 == expected, j2
+  let j3 = j2.parseJson
+  let t2 = j3.jsonTo(T)
+  doAssert t2 == t # xxx handle NaN specially, using isNaN once https://github.com/nim-lang/Nim/issues/18007 is fixed
+  doAssert $t2.toJson == j2 # still needed, because -0.0 = 0.0 but their json representation differs
+
+proc testRoundtrip2[T](t: T, expected: string) =
```

Comment:
redundant with `testRoundtripVal`, will remove this in next PR or this PR if there are other things to address

Reply:
done